app/src/main/java/com/github/sdpteam15/polyevents/database/observe/ObservableList.kt

```diff
@@ -48,85 +44,191 @@ class ObservableList<T> {
+                                it.value,
+                                values.indexOf(observable)
+                            ), it.sender
+                        )
+                    )
+                }
+            itemAdded(UpdateArgs(observable.value!!, sender))
+            notifyUpdate(sender)
+            return observable
+        }
+        return null
     }
 
     /**
      * Add an item.
-     * @param item item to add.
-     * @return observable added.
+     * @param item Item to add.
+     * @param sender The source of the event.
+     * @return Observable added.
      */
-    fun add(item: T): Observable<T> = add(Observable(item))
+    fun add(item: T, sender: Any? = null): Observable<T>? = add(Observable(item), sender, true)!!
+    private fun add(item: T, sender: Any? = null, notify: Boolean): Observable<T>? =
+        add(Observable(item), sender, notify)!!
+
+    override fun addAll(elements: Collection<T>): Boolean = addAll(elements, null)
+
+    /**
+     * Add all items in the list.
+     * @param items items list.
+     * @param sender The source of the event.
+     */
+    fun addAll(items: Collection<T>, sender: Any? = null): Boolean {
+        var res = true
+        for (item: T in items)
+            res = res && (add(item, sender, false) != null)
+        notifyUpdate(sender)
+        return res
+    }
+
+    override fun remove(element: T): Boolean = remove(element, null) != null
 
     /**
      * Remove an item.
-     * @param item item to remove.
+     * @param observable item to remove.
+     * @param sender The source of the event.
      * @return observable removed.
      */
-    fun remove(observable: Observable<T>): Observable<T>? {
+    fun remove(observable: Observable<T>, sender: Any? = null): Observable<T>? =
+        remove(observable, sender, true)
+
+    private fun remove(
+        observable: Observable<T>,
+        sender: Any? = null,
+        notify: Boolean
+    ): Observable<T>? {
         if (!values.remove(observable))
             return null
         removeItemObserver[observable]!!()
         removeItemObserver.remove(observable)
-        itemRemoved(observable.value)
+        itemRemoved(UpdateArgs(observable.value!!, sender))
+        if (notify)
+            notifyUpdate(sender)
         return observable
     }
 
     /**
      * Remove an item.
      * @param item item to remove.
+     * @param sender The source of the event.
      * @return observable removed.
      */
-    fun remove(item: T): Observable<T>? {
-        val observable: Observable<T>? = values.find { it.value == item }
+    fun remove(item: T, sender: Any? = null): Observable<T>? = remove(item, sender, true)
+    private fun remove(item: T, sender: Any? = null, notify: Boolean): Observable<T>? {
+        val observable: Observable<T>? = find(item)
         if (observable != null)
-            return remove(observable)
+            return remove(observable, sender, notify)
```

Comment:
I would recommend using the brackets when writing an if.
If you go modify the code later, you might forget to add them if you want to add a line in this if

Reply:
Same for all if, and for

Reply:
Yeah this could easily get us stuck 1 hour because we add for example a Log.d just above. Changed in next commit !